include/seqan3/alignment/matrix/detail/alignment_trace_matrix_full.hpp

```diff
@@ -0,0 +1,202 @@
+// -----------------------------------------------------------------------------------------------------
+// Copyright (c) 2006-2019, Knut Reinert & Freie Universität Berlin
+// Copyright (c) 2016-2019, Knut Reinert & MPI für molekulare Genetik
+// This file may be used, modified and/or redistributed under the terms of the 3-clause BSD-License
+// shipped with this file and also available at: https://github.com/seqan/seqan3/blob/master/LICENSE.md
+// -----------------------------------------------------------------------------------------------------
+
+/*!\file
+ * \brief Provides seqan3::detail::alignment_trace_matrix_full.
+ * \author Rene Rahn <rene.rahn AT fu-berlin.de>
+ */
+
+#pragma once
+
+#include <seqan3/alignment/matrix/detail/alignment_matrix_column_major_range.hpp>
+#include <seqan3/alignment/matrix/detail/alignment_trace_matrix_base.hpp>
+#include <seqan3/alignment/matrix/detail/alignment_trace_matrix_proxy_type.hpp>
+#include <seqan3/std/iterator>
+#include <seqan3/std/ranges>
+
+namespace seqan3::detail
+{
+
+/*!\brief An alignment traceback matrix storing the entire traceback matrix.
+ * \tparam trace_t          The type of the trace directions.
+ * \tparam coordinate_only  A boolean flag indicating if only a seqan3::alignment_coordinate should be generated.
+ * \ingroup alignment_matrix
+ *
+ * \details
+ *
+ * This implementation allocates the full traceback matrix using quadratic memory. The matrix allows access to the
+ * underlying values through a range based interface. An iterator over the traceback matrix iterates in
+ * column-major-order over the traceback matrix. Dereferencing an iterator returns a view over the current matrix
+ * column. The value type is a pair over a seqan3::alignment_coordinate and
+ * the seqan3::detail::alignment_trace_matrix_proxy_type, which gives a unified access to the respective matrix cells
+ * as needed by the standard alignment algorithm. The matrix is modelled as std::ranges::InputRange since the
+ * alignment algorithm iterates only once over the complete matrix to calculate the values.
+ *
+ * ### Only computing the coordinates
+ *
+ * Sometimes it is desired to only get access to the alignment coordinates. This can be achieved by setting
+ * `coordinate_only = true`. In this case no memory will be allocated and only an internal state is maintained to
+ * generate the alignment coordinates.
+ */
+template <typename trace_t, bool coordinate_only = false>
+class alignment_trace_matrix_full :
+    public alignment_trace_matrix_base<trace_t>,
+    public alignment_matrix_column_major_range<alignment_trace_matrix_full<trace_t, coordinate_only>>
+{
+private:
+
+    static_assert(std::Same<trace_t, trace_directions> || Simd<trace_t>,
+                  "Value type must either be a trace_directions object or a Simd vector.");
```

Comment:
That won't be possible since the vector type can only store regular integer types. It would be possible only to check at runtime if the value does no exceed the values of trace_directions, but of course you don't want to do that. Also in an assert you don't.